ArmPkg/Library/ArmArchTimerLib/ArmArchTimerLib.c

```diff
@@ -126,13 +128,16 @@ MicroSecondDelay (
                    );
 
   // Read System Counter value
-  SystemCounterVal = ArmGenericTimerGetSystemCount ();
-
-  TimerTicks64 += SystemCounterVal;
+  PreviousSystemCounterVal = ArmGenericTimerGetSystemCount ();
 
   // Wait until delay count expires.
-  while (SystemCounterVal < TimerTicks64) {
+  while (TimerTicks64 > 0) {
     SystemCounterVal = ArmGenericTimerGetSystemCount ();
+    // Get how much we advanced this tick. Wrap around still has delta correct
+    DeltaCounterVal = SystemCounterVal - PreviousSystemCounterVal;

```

Comment:
If the timer is only 56 bits wide, we might wrap here from 2^56-1 to 0, producing a delta with a very high value. That would result in the delay loop to complete prematurely, which is exactly the behavior we are intending to fix here.

```
DeltaCounterVal &= MAX_UINT64 >> 8 // Timer may be only 56 bits wide
```

should take care of this I think.

Reply:
aah yeah - i did the code before i saw your 56bit comment. i did not go look up the ARM ARM :)